staging/kos/cmd/controllers/connectionagent/main.go

```diff
@@ -0,0 +1,171 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"flag"
+	"os"
+	"os/signal"
+	"runtime"
+	"syscall"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/workqueue"
+
+	kosclientset "k8s.io/examples/staging/kos/pkg/client/clientset/versioned"
+	cactlr "k8s.io/examples/staging/kos/pkg/controllers/connectionagent"
+	netfabricfactory "k8s.io/examples/staging/kos/pkg/networkfabric/factory"
+)
+
+const (
+	nodeNameEnv      = "NODE_NAME"
+	hostIPEnv        = "HOST_IP"
+	netFabricTypeEnv = "NET_FABRIC_TYPE"
+
+	defaultNumWorkers  = 2
+	defaultClientQPS   = 100
+	defaultClientBurst = 200
+)
+
+func main() {
+	var (
+		nodeName               string
+		hostIP                 string
+		netFabricType          string
+		kubeconfigFilename     string
+		workers                int
+		clientQPS, clientBurst int
+	)
+	flag.StringVar(&nodeName, "nodename", "", "node name")
+	flag.StringVar(&hostIP, "hostip", "", "host IP")
+	flag.StringVar(&netFabricType, "netfabric", "", "network fabric type")
+	flag.StringVar(&kubeconfigFilename, "kubeconfig", "", "kubeconfig filename")
+	flag.IntVar(&workers, "workers", defaultNumWorkers, "number of worker threads")
+	flag.IntVar(&clientQPS, "qps", defaultClientQPS, "limit on rate of calls to api-server")
+	flag.IntVar(&clientBurst, "burst", defaultClientBurst, "allowance for transient burst of calls to api-server")
+	flag.Set("logtostderr", "true")
+	flag.Parse()
+
+	if len(os.Getenv("GOMAXPROCS")) == 0 {
+		runtime.GOMAXPROCS(runtime.NumCPU())
+	}
+
+	nodeName, err := getNodeName(nodeName)
+	if nodeName == "" {
+		glog.Errorf("Could not retrieve node name: neither command line flag \"nodename\" nor env var \"NODE_NAME\" were set. Lookup of OS hostname failed: %s\n",
+			err.Error())
+		os.Exit(2)
+	}
+
+	hostIP = getHostIP(hostIP)
+	if hostIP == "" {
+		glog.Errorf("Could not retrieve host IP: neither command line flag \"hostip\" nor env var \"HOST_IP\" were provided\n")
+		os.Exit(3)
+	}
+
+	netFabricType = getNetFabricType(netFabricType)
+	// No check on whether netFabricType is set because the
+	// factory falls back to a default implementation if not.
+	netFabric := netfabricfactory.NewNetFabricForType(netFabricType)
+
+	clientCfg, err := clientcmd.BuildConfigFromFlags("", kubeconfigFilename)
+	if err != nil {
+		glog.Errorf("Failed to build client config for kubeconfig=%q: %s\n", kubeconfigFilename, err.Error())
+		os.Exit(4)
+	}
+	clientCfg.QPS = float32(clientQPS)
+	clientCfg.Burst = clientBurst
+
+	kcs, err := kosclientset.NewForConfig(clientCfg)
+	if err != nil {
+		glog.Errorf("Failed to build KOS clientset for kubeconfig=%q: %s\n", kubeconfigFilename, err.Error())
+		os.Exit(5)
+	}
+
+	// TODO think whether the rate limiter parameters make sense
+	queue := workqueue.NewRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(200*time.Millisecond, 8*time.Hour))
```

Comment:
Make it a named queue so that it will produce Prometheus metrics.

Reply:
Added named queue in 06d37b6.